### ISSUE TITLE
How do I disable auto-renew without removing config files?

### ISSUE THREAD

**Issue by ethaniel** [certbot/certbot#8555]
The new snapd system is absolutely nuts. It keeps calling the certbot auto-renewal process every 12 (or so) hours without allowing me to turn it off in any way.

The reason is that I want that feature turned off is that because I have my own layer on top of certbot (merging certs into one file, putting that file on shared storage, sending a signal to frontends to read the new certs).

And besides, I don't feel comfortable with something living on my servers on it's own without my control.
How do I disable "snap.certbot.renew.service"?

**Comment by alexzorin** [comment]
The Certbot snap works similarly to the Debian package. They both come with a systemd timer that calls `certbot renew -q` twice a day.

If you having some kind of orchestration system on top of Certbot, then you should be able to disable the scheduled task:

    sudo snap stop --disable certbot.renew

Only thing, in my attempts to verify, this doesn't seem to actually work. The timer still triggers the service and the service still launches Certbot.

Disabling `snap.certbot.renew.{timer,service}` via `systemctl` doesn't seem like an option either, because upon refresh of the snap, it goes back to an enabled state.

See also [this thread on snapcraft.io](https://forum.snapcraft.io/t/how-to-disable-a-timer-service-in-a-snap/7602/2).

I am tentatively going to mark this as a bug, pending a closer investigation into how this is supposed to work.

**Comment by basak** [comment]
> I am tentatively going to mark this as a bug,

I think this is tracked at https://bugs.launchpad.net/snapd/+bug/1842258

**Comment by p-schneider** [comment]
I think this was fixed in snap version 2.49 (released in February)

**Comment by runthis** [comment]
With certbot auto-renew causing nginx to fail every 60 days, it would be convenient to disable auto-renew.

**Comment by sheodox** [comment]
Just found this issue because I was having the same problem with my sites all going down in the middle of the night every couple months. I switched from my own automation to making pre/post hook scripts for certbot and judging by the output when running `certbot renew --force-renewal` (certs just renewed and took down my sites last night, so I had brand new certs) it looks to be working for my needs.

From the docs: https://eff-certbot.readthedocs.io/en/stable/using.html?highlight=hooks#setting-up-automated-renewal

> If you needed to stop your webserver to run Certbot, you’ll want to add pre and post hooks to stop and start your webserver automatically. For example, if your webserver is HAProxy, run the following commands to create the hook files in the appropriate directory:
> 
> ```
> sudo sh -c 'printf "#!/bin/sh\nservice haproxy stop\n" > /etc/letsencrypt/renewal-hooks/pre/haproxy.sh'
> sudo sh -c 'printf "#!/bin/sh\nservice haproxy start\n" > /etc/letsencrypt/renewal-hooks/post/haproxy.sh'
> sudo chmod 755 /etc/letsencrypt/renewal-hooks/pre/haproxy.sh
> sudo chmod 755 /etc/letsencrypt/renewal-hooks/post/haproxy.sh
> ```
> 
> Congratulations, Certbot will now automatically renew your certificates in the background.

My pre hook just stops haproxy, post hook concats the keys together so haproxy can read them then starts haproxy.

**Comment by bwesen** [comment]
Hi, was this ever fixed? Is it possible to disable auto renew properly? I just found 999 log files filled with errors from snap-certbot running by itself every 12 hours :) (I just want to renew when I tell certbot to do it as I need to do DNS verification and I don't want to put the DNS changing scripts on the same server).

**Comment by alexzorin** [comment]
@bwesen I think you can probably do this with a systemd override file.

    systemctl edit snap.certbot.renew.service

making the contents: 

    [Timer]
    OnCalendar=

followed by:

    systemctl daemon-reload

This should persist through snap upgrades to never run the timer.

**Comment by fedkad** [comment]
The `daemon-reload` command logs a message like

```
snap.certbot.renew.timer: Timer unit lacks value setting. Refusing.
```

 to system log. Is this OK?

**Comment by alexzorin** [comment]
@fedkad yes, we are intentionally making the timer configuration invalid, so it doesn't get loaded. That's what the warning is about.

**Comment by alexzorin** [comment]
I don't have time to test this right now, but I believe you need to do this:

    [Timer]
    OnCalendar=
    OnCalendar=--* 01:00:00

In order to clear the existing value. Otherwise, you both `OnCalendar` values take effect.

**Comment by VietDuc19** [comment]
Thank you so much for your help @alexzorin, it worked. I also change renewal time to other time, not at the top of hours. Hope you can fix it soon.

**Comment by Spudz76** [comment]
Wouldn't masking it work no matter what?

`systemctl mask snap.certbot.renew.timer`

and perhaps to be extra sure

`systemctl mask snap.certbot.renew.service`

since the timer just starts the service, but a masked service can't start, so even if the timer tries to happen for any reason it still can't run the service

**Comment by alexzorin** [comment]
I think masking, which re-links the units to `/dev/null`, is only possible for units that are symlinks into `/lib/systemd/system`. The units created in `/etc/systemd/system` by snapd are real files, not symlinks:

    # systemctl mask snap.certbot.renew.timer
    Failed to mask unit: File /etc/systemd/system/snap.certbot.renew.timer already exists.

**Comment by Spudz76** [comment]
[EXPLETIVE RAGE REDACTED, but do know I used ALL of the words]

grrrrrrrrr... junk is never supposed to make files in `/etc/systemd/` precisely because it is incompatible with masking.  EVERYTHING puts its stuff in `/lib/systemd/` whether it's add-on or core or user-junk or not

So either move the file(s) to where they [EXPLETIVES, LOUDLY] belong, until upstream fixes their idiocy, and then masking will work.  But then I bet the next crapd (oops, not oops) update will delete the mask links and put their timer virus back, in the wrong place again.

**Comment by RobertJHowell** [comment]
Thank you @alexzorin
I disabled auto-renew without removing config files on my Ubuntu 20.4 server using your method :

sudo systemctl edit snap.certbot.renew.timer
I entered 2 blank timer lines as follows
[Timer]
OnCalendar=
OnCalendar=

The editing operation put these override lines in file cat /etc/systemd/system/snap.certbot.renew.timer.d/override.conf

Then restarted systemctl with systemctl daemon-reload
Then verified certbot service (snap.certbot.renew.service) wasn't in the list of active timers with sudo systemctl list-timers

**Comment by neillbell** [comment]
I'm running the certbot snap version 2.11.0.

I thought that I had this sorted using the hack for the renew timer above, but I just discovered that one of my certs was updated outside of my knowledge. Unfortunately, my syslogs don't go back far enough to see if the systemd snap.certbot.renew.service was triggered somehow.

Is the renew timer hack the proper way to disable automatic updates, or does the snap provide something else?

**Comment by neillbell** [comment]
Disabling the timer was the first thing I tried, but it eventually gets re-enabled somehow. I tried the OnCalendar trick mentioned above, but that didn't work either.

I eventually settled on the [pip version](https://certbot.eff.org/instructions?ws=apache&os=pip) mentioned on the certbot website. It doesn't do anything sneaky behind your back.

**Comment by igorpupkinable** [comment]
> Has anyone tried `sudo systemctl disable snap.certbot.renew.timer`? It would seem that this is the proper way to disable the timer. Or does it somehow come back?

[Snap fix](https://github.com/canonical/snapd/pull/13128) has been released in version 2.61. So it may work with `sudo snap stop --disable certbot.renew` properly.
